### PR TITLE
Fix #905. Implemented user manager plugin

### LINKED ISSUE #905: Implement User Management as per MapStore 1
See mockups below:

1. [user manager](http://lf.mapstore2.geo-solutions.it/samples/map/manager.html)
2. [add user 1](http://lf.mapstore2.geo-solutions.it/samples/map/add-user-1.html)
3. [add user 2](http://lf.mapstore2.geo-solutions.it/samples/map/add-user-2.html)
4. [add user 3](http://lf.mapstore2.geo-solutions.it/samples/map/add-user-3.html)

### PR DESCRIPTION
**Pull request by offtherailz**
Implemented user manager.

image

* Creation / Editing /Deletion of users

image
image
image

* Minimal support to group management (no creation or delete, only add to the user).
image
image

* Plus: 
    * Improvements to dialog (support loading mask and modal mode)
    * Add a reusable confirm dialog already configured.

### CODE REVIEW

File: web/client/actions/users.js [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                if (userLoaded && userLoaded.groups === "") {
+                    userLoaded = {...userLoaded, groups: user.groups};
+                }
+                dispatch({
+                    type: USERMANAGER_EDIT_USER,
+                    status: "success",
+                    user: userLoaded
+                });
+            }).catch((error) => {
+                dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+                    type: USERMANAGER_EDIT_USER,
+                    status: "error",
+                    user,
+                    error
+                });
+            });
+        }
+        dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+            type: USERMANAGER_EDIT_USER,
+            user: user
+        });
+    };
+}
+function saveUser(user, options = {}) {
+    return (dispatch) => {
+        if (user && user.id) {
+            dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                type: USERMANAGER_UPDATE_USER,
+                status: "saving",
+                user
+            });
+            return API.updateUser(user.id, {...user, group: user.groups}, options).then((userDetails) => {
+                dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+                    type: USERMANAGER_UPDATE_USER,
+                    status: "saved",
+                    user: userDetails && userDetails.User
+                });
+            }).catch((error) => {
+                dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+                    type: USERMANAGER_UPDATE_USER,
+                    status: "error",
+                    user,
+                    error
+                });
+            });
+        }
+        // createUser
+        dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+            type: USERMANAGER_UPDATE_USER,
+            status: "creating",
+            user
+        });
+        return API.createUser(user, options).then((id) => {
+            dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+                type: USERMANAGER_UPDATE_USER,
+                status: "created",
+                user: { ...user, id}
+            });
+            dispatch(getUsers());
+        }).catch((error) => {
+            dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+function deleteUser(id, status = "confirm") {
+    if (status === "confirm" || status === "cancelled") {
+        return {
+            type: USERMANAGER_DELETE_USER,
+            status,
+            id
+        };
+    } else if ( status === "delete") {
+        return (dispatch) => {
+            dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+                type: USERMANAGER_DELETE_USER,
+                status: "deleting",
+                id
+            });
+            API.deleteUser(id).then(() => {
+                dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/actions/users.js [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+                    type: USERMANAGER_DELETE_USER,
+                    status: "deleted",
+                    id
+                });
+                dispatch(getUsers());
+            }).catch((error) => {
+                dispatch({
```
Comment (mbarto) [inline web/client/actions/users.js]: define an action creator, please

File: web/client/components/manager/users/UserGroups.jsx [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    onChange(values) {
+        if (values === null) {
+            return;
+        }
+        this.props.onUserGroupsChange("groups", values.map((group) => {
+            let index = this.props.groups.findIndex((availableGroup)=>availableGroup.id === group.value);
+            return index >= 0 ? this.props.groups[index] : null;
+        }).filter(group => group));
+    },
+    getEveryOneGroups() {
```
Comment (mbarto) [inline web/client/components/manager/users/UserGroups.jsx]: getEveryOneGroup

File: web/client/localConfig.json [hunk 13]
```diff
@@ -737,18 +747,31 @@
 
 		"maps": [
 			{"name": "Header"},
-			{"name": "Fork"},
```
Comment (mbarto) [inline web/client/localConfig.json]: Why did you remove Fork and Language?

File: web/client/plugins/manager/users/UserDialog.jsx [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+const {bindActionCreators} = require('redux');
+const {connect} = require('react-redux');
+const {editUser, changeUserMetadata, saveUser} = require('../../../actions/users');
+
+
+const mapStateToProps = (state) => {
```
Comment (mbarto) [inline web/client/plugins/manager/users/UserDialog.jsx]: Do we need to make it so complex? Please, use the usual selectors composition with reselect

File: web/client/plugins/manager/users/UserDialog.jsx [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    const users = state && state.users;
+    return {
+        modal: true,
+        show: users && !!users.currentUser,
+        user: users && users.currentUser,
+        groups: users && users.groups
+    };
+};
+const mapDispatchToProps = (dispatch) => {
```
Comment (mbarto) [inline web/client/plugins/manager/users/UserDialog.jsx]: bindActionCreators is not needed, bind is not needed

File: web/client/plugins/manager/users/UserGrid.jsx [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    };
+};
+const mapDispatchToProps = (dispatch) => {
+    return bindActionCreators({
+        loadUsers: getUsers,
+        onEdit: editUser,
+        onDelete: deleteUser
+    }, dispatch);
+};
+const mergeProps = (stateProps, dispatchProps, ownProps) => {
```
Comment (mbarto) [inline web/client/plugins/manager/users/UserGrid.jsx]: Not clear the purpose of this...

File: web/client/reducers/users.js [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+ */
+
+const {
+    USERMANAGER_GETUSERS, USERMANAGER_EDIT_USER, USERMANAGER_EDIT_USER_DATA, USERMANAGER_UPDATE_USER, USERMANAGER_DELETE_USER,
+    USERMANAGER_GETGROUPS
+} = require('../actions/users');
+const assign = require('object-assign');
+function users(state = {
+    start: 0,
+    limit: 12
```
Comment (mbarto) [inline web/client/reducers/users.js]: 12 is duplicated everywhere, can we make this a constant that we change in a single place?

File: web/client/components/manager/users/UserGrid.jsx [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+                verticalAlign: "center",
+                left: "0",
+                background: "rgba(255, 255, 255, 0.5)",
+                zIndex: 2
+            }}><div style={{
+                  position: "absolute",
+                  top: "50%",
+                  left: "50%",
+                  transform: "translate(-50%, -40%)"
+            }}>Loading...<Spinner spinnerName="circle" noFadeIn/></div></div>);
```
Comment (mbarto) [inline web/client/components/manager/users/UserGrid.jsx]: This should be localized

File: web/client/localConfig.json (hunk 13 above)
Comment (offtherailz) [inline web/client/localConfig.json]: Something with merge with master went wrong, going to recover it

File: web/client/components/manager/users/UserGroups.jsx (hunk 12 above)
Comment (offtherailz) [inline web/client/components/manager/users/UserGroups.jsx]: It returns an array with groups, should I call it getEveryOneGroupArray?

Comment (mbarto) [inline web/client/components/manager/users/UserGroups.jsx]: With that name it seems you can have many everyone groups? Is it true, or you always get an array of exactly one element?

File: web/client/plugins/manager/users/UserDialog.jsx (hunk 15 above)
Comment (offtherailz) [inline web/client/plugins/manager/users/UserDialog.jsx]: line 25, bind the action with null as first parameter, this is needed.

File: web/client/plugins/manager/users/UserDialog.jsx (hunk 14 above)
Comment (offtherailz) [inline web/client/plugins/manager/users/UserDialog.jsx]: They are only not null checks, how it is less complex and more readable than this?

File: web/client/plugins/manager/users/UserGrid.jsx (hunk 16 above)
Comment (offtherailz) [inline web/client/plugins/manager/users/UserGrid.jsx]: allow overriding from ownProps , start and limit , even if not defined.

File: web/client/actions/generators-importer.js [hunk 19]
```diff
@@ -0,0 +1,615 @@
+/**
```
Comment (offtherailz) [inline web/client/actions/generators-importer.js]: going to remove them

File: web/client/actions/users.js [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+    let start;
+    let limit;
+    if (params) {
+        start = params.start;
+        limit = params.limit;
+    }
+    return (dispatch, getState) => {
+        let text = searchText;
+        let state = getState && getState();
+        if (state) {
```
Comment (offtherailz) [inline web/client/actions/users.js]: Yes I know that search text and current page should come from the action, but this way the component are not indipendent.  Where you should place them

File: web/client/components/manager/users/UserGroups.jsx (hunk 12 above)
Comment (offtherailz) [inline web/client/components/manager/users/UserGroups.jsx]: They are default groups. Going to replace it with `getDefaultGroups`

File: web/client/actions/users.js [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+            let totalCount = get(response, "ExtUserList.UserCount");
+            users = Array.isArray(users) ? users : [users];
+            users = users.map((user) => {
+                let groups = get(user, "groups.group");
+                groups = Array.isArray(groups) ? groups : [groups];
+                return assign({}, user, {
+                    groups
+                });
+            });
+            dispatch({
```
Comment (offtherailz) [inline web/client/actions/users.js]: Didn't find any convention that forces to create module internal function that create the final object from the action creator. I thought to improve readability in this way (avoid to jump here and there for simple objects definition, avoid passing variables that can be inverted and so on...) I fixed this, let's check if it is a good convention and lets write it if there is a reason to make it mandatory.

Comment (mbarto) [inline web/client/actions/users.js]: The only reason for now is: we always did this way, so far, let's discuss if we want to improve it and how, but I agree it's not mandatory

File: web/client/plugins/manager/users/UserDialog.jsx (hunk 15 above)
Comment (mbarto) [inline web/client/plugins/manager/users/UserDialog.jsx]: ok, got it, do we also need it for onChange and onSave ?

File: web/client/plugins/manager/users/UserGrid.jsx (hunk 16 above)
Comment (mbarto) [inline web/client/plugins/manager/users/UserGrid.jsx]: ok
